src/hugr/serialize.rs

```diff
@@ -89,120 +95,99 @@ impl<'de> Deserialize<'de> for Hugr {
+                    assert_eq!(hugr.root(), n);
                     n
                 });
-                let opt = &op_types[n];
-                Ok((
-                    parent,
-                    graph.num_inputs(n),
-                    graph.num_outputs(n),
-                    opt.clone(),
-                ))
+                let opt = hugr.get_optype(n);
+                Ok((parent, opt.clone()))
             })
             .collect::<Result<_, Self::Error>>()?;
-        for (parent, _, _, _) in &mut nodes {
+        for (parent, _) in &mut nodes {
             *parent = node_rekey[parent];
         }
 
-        let find_offset = |node: NodeIndex, offset: usize, dir: Direction| {
-            let sig = &op_types[node].signature();
+        let find_offset = |node: Node, offset: usize, dir: Direction, hugr: &Hugr| {
+            let sig = hugr.get_optype(node).signature();
             let offset = match offset < sig.port_count(dir) {
                 true => Some(offset as u16),
                 false => None,
             };
             (node, offset)
         };
 
-        let edges: Vec<_> = graph
-            .nodes_iter()
+        let edges: Vec<_> = hugr
+            .nodes()
             .flat_map(|node| {
-                graph
-                    .outputs(node)
+                hugr.node_ports(node, Direction::Outgoing)
                     .enumerate()
                     .flat_map(move |(src_offset, port)| {
-                        let src = find_offset(node, src_offset, Direction::Outgoing);
-                        graph.port_links(port).map(move |(_, tgt)| {
-                            let tgt_node = graph.port_node(tgt).unwrap();
-                            let tgt_offset = graph.port_offset(tgt).unwrap().index();
-                            let tgt = find_offset(tgt_node, tgt_offset, Direction::Incoming);
+                        let src = find_offset(node, src_offset, Direction::Outgoing, hugr);
+                        hugr.linked_ports(node, port).map(move |(tgt_node, tgt)| {
+                            let tgt = find_offset(
+                                tgt_node,
+                                tgt.offset.index(),
+                                Direction::Incoming,
+                                hugr,
+                            );
                             [src, tgt]
                         })
                     })
             })
             .collect();
 
-        Ok(Self {
-            nodes,
-            edges,
-            root: *root,
-        })
+        Ok(Self { nodes, edges })
     }
 }
 
 impl TryFrom<SerHugrV0> for Hugr {
     type Error = HUGRSerializationError;
-    fn try_from(SerHugrV0 { nodes, edges, root }: SerHugrV0) -> Result<Self, Self::Error> {
-        let mut hierarchy = Hierarchy::new();
-
+    fn try_from(SerHugrV0 { nodes, edges }: SerHugrV0) -> Result<Self, Self::Error> {
+        // Root must be first node
```

Comment:
We don't explicitly require this anywhere else. We should either:

- State this as a property of the Hugr, checked during validation.
- Allow `root != 0`, and could logically swap `NodeIndex(0)` and `root` during the encoding

Reply:
option 2: this would mean root index isn't preserved, but that's ok because indices get messed about anyway? 
I think option 2 sounds good

Reply:
done